src/nupic/regions/SPRegion.py

```diff
@@ -844,6 +851,54 @@ def setParameter(self, parameterName, index, parameterValue):
   #############################################################################
 
 
+  def write(self, proto):
+    """Write state to proto object.
+
+    proto: PyRegionProto capnproto object
+    """
+    regionImpl = proto.regionImpl.as_struct(SPRegionProto)
+
+    regionImpl.spatialImp = self.spatialImp
+    regionImpl.columnCount = self.columnCount
+    regionImpl.inputWidth = self.inputWidth
+    regionImpl.learningMode = 1 if self.learningMode else 0
+    regionImpl.inferenceMode = 1 if self.inferenceMode else 0
+    regionImpl.anomalyMode = 1 if self.anomalyMode else 0
+    regionImpl.topDownMode = 1 if self.topDownMode else 0
+
+    self._sfdr.write(regionImpl.spatialPooler)
+
+
+  @classmethod
+  def read(cls, proto):
+    """Read state from proto object.
+
+    proto: PyRegionProto capnproto object
+    """
+    regionImpl = proto.regionImpl.as_struct(SPRegionProto)
+
+    instance = cls(regionImpl.columnCount, regionImpl.inputWidth)
+
+    instance.spatialImp = regionImpl.spatialImp
+    instance.learningMode = regionImpl.learningMode
+    instance.inferenceMode = regionImpl.inferenceMode
+    instance.anomalyMode = regionImpl.anomalyMode
+    instance.topDownMode = regionImpl.topDownMode
+
+    spatialImp = regionImpl.spatialImp
+
+    if spatialImp == 'py':
+      instance._sfdr = PYSpatialPooler.read(regionImpl.spatialPooler)
+    elif spatialImp == 'cpp':
+      instance._sfdr = CPPSpatialPooler()
```

Comment:
We should make a constructor for the C++ implementation that takes the proto instance rather than using the empty constructor and then `read`.